### ISSUE TITLE
AI of Deepmind discovered 4x4 x 4x4 non-comm. multip. algorithm with just 47 mul.

### ISSUE THREAD

**Issue by ValZapod** [xianyi/OpenBLAS#3783]
https://www.nature.com/articles/s41586-022-05172-4

This of course will destroy all sceptics who did not believe that neural networks can do math in symbolic or even pure form. Considering oneMKL is only so fast because of libxsmm library (and by itself outperfoms everything else) I will report it [here](https://github.com/libxsmm/libxsmm/issues/698) first.
THIS ALGORITHM is **non-commutative**, i.e. has A elements on one side and B matrix elements on the other of EACH "*". THAT means it can be used recursively to multiply block matrices.

![41586_2022_5172_Fig6_ESM](https://user-images.githubusercontent.com/31514790/194192327-bb3eef24-5464-430c-98d8-47bfdfcbf529.jpg)

**Comment by brada4** [comment]
It is significantly more flops per point, and n^3 temporary space. 
BLAS supposedly does full evaluation,i.e 2 muls one add per point, fitting temp vars into whatever register file there is.
You can find more of those: O(?) is good, but linear slowdown is so wild you never profit in practice.

**Comment by martin-frbg** [comment]
Thanks for the pointer, but on first impression I'm agreeing with brada4's comment and also those raised in https://github.com/libxsmm/libxsmm/issues/698 where you cross-posted. Certainly interesting than a neural network can be trained to generate such algorithms, but the practical value (beyond getting published in Nature) remains to be seen. (If you or somebody else _does_ come up with a demonstrably faster implementation based on this, a PR will certainly be welcome, but it just does not look tempting "on paper")

**Comment by ValZapod** [comment]
Apparently there are some results that work even in not modular arithemtic. See this 47 result is only imporoved in modular, but 3, 4, 5 is improved in both. This is novel result.

![20221006_112322](https://user-images.githubusercontent.com/31514790/194261146-0bcfad2d-1f40-44b7-96c7-3f9b247905ac.jpg)

**Comment by ValZapod** [comment]
>It is not 486DX where MUL was 10x slower than ADD

Mmm, it still requires more electricty. Just as new divq on Ice lake, 128 bit division. Or M1 division, that is somehow even fasterm

**Comment by brada4** [comment]
FMA can be done on packed ?MM registers, scattered ADD not really, but for sake of completeness implementation is nice to have.

**Comment by brada4** [comment]
Lets take c11 where a32 and a34 are introduced via h15 but never subtracted. Actually nothing is ever subtracted 
If you have it in text format try to symbolically solve in sympy or maxima.

**Comment by brada4** [comment]
Lets multiply 4x4 identity matrix using algorithm.....  h2==4 , no matter how many zeroes and ones you add there - you never get 1.0

**Comment by ValZapod** [comment]
>If you have it in text format try to symbolically solve in sympy or maxima.

I do not, but there is code that does print that? https://github.com/deepmind/alphatensor/issues/3

Also this 4, 4, 4 result is for mod-2 tensors.

**Comment by ValZapod** [comment]
All results. Both modular and normal arithmetic. 7 MB file. 
[equations.txt](https://github.com/libxsmm/libxsmm/files/9737591/equations.txt)

**Comment by brada4** [comment]
mod-2 - are you making fun? OP mentions general purpose float libraries, but you propose integer parity algorithm.

**Comment by ValZapod** [comment]
I am OP... Original poster. And they are no all mod-2.

**Comment by brada4** [comment]
I am not a tensor, hard to miss obvious disintegrity. modulo-2 4x4 "matrix" is historically known as uint16.

which is the 4x4 optimal stuff considering cost of mul is same as add/sub nowadays?

**Comment by ValZapod** [comment]
>which is the 4x4 optimal stuff considering cost of mul is same as add/sub nowadays?

It is not on ucode level. That is physically impossible. You can measure uinstructions and electricity.

**Comment by ValZapod** [comment]
>I am not a tensor

🤔 ?

**Comment by brada4** [comment]
i have internet access. that does not make your demagogical attempt at pushing unfit algorithm more valid.

**Comment by ValZapod** [comment]
Google said it helped their silicon already. So it was fit for HW already. Also see https://github.com/paolodalberto/FastMMW/issues/1

**Comment by ValZapod** [comment]
>i have internet access.

I do to. You should be in Github's AS to be able to converse here and not be connected to the Internet.

**Comment by brada4** [comment]
The problem of multiple adds is precision, if the unrelated part that is algebraically correctly  added and subtracted is significantly larger (more than e7 apart) than the actual value you basically lose all precision for that point. In the normal algorithm it goes standard way so that precision is cancelled once per output value.

Another for parity aka modulo2 - if you degrade inputs to parity (single AND with mask), all multiplications become AND and all additions XOR, so one can claim it's complexity is 0(zero) because there is no multiplication whatsoever.

In modern computers MUL is twice as expensive as ADDSUB, so in reality 4x4 is not an optimisation when you trade one MUL for ?N? ADD-s.
But it has a chance in mangling blocks out of order. Thats in the paper too - Straessen's application in benchmarks, that does not claim breakthrough and is taken as something obvious. Still the rounding defect propagates BUT mul is N^3 but add is N^2.

**Comment by ValZapod** [comment]
BTW, using those results (5, 5, 5) was improved to 97 in normal arithmetic and to 95 in mod 2. Non-commutative, so immediately useful! 
https://arxiv.org/abs/2212.01175

Are you sure you do not want this result with commutative (3, 3, 3) in just 21 operations?
https://arxiv.org/abs/1904.07683 (published: https://www.sciencedirect.com/science/article/abs/pii/S0747717122000499)

I even wrote code for 21 (3, 3, 3): 
https://gcc.godbolt.org/z/9oszvM9aG

**Comment by brada4** [comment]
add vs mul is still 1:2 uops and at this small scale few muls less for ten adds is significantly worse. I suppose the code grows exponentially as matrices grow, so it runs out of realistic L1i cache and makes L2 flush out and resort to 100+uops load/store to main RAM.

You need to compile your gcc with `-O2 -fno-tree-vectorize` as the latest seems to over-read inputs and make like 10x longer code that dynamically attempts ?some alignment heuristic?

BUT findings are valid to design a silicon that does mmul a tiny hair faster...

**Comment by brada4** [comment]
Basically compare code to respective  [netlib](https://netlib.org/lapack/explore-3.1.1-html/dgemm.f.html) reference code in same place.
